src/grafana_service.ts

```diff
@@ -26,6 +26,77 @@ const CHUNK_SIZE = 50000;
  * @param metric to query to Grafana
  * @returns { values: [time, value][], columns: string[] }
  */
+
+export class DatasourceRequest {
+
+  constructor(public metric: Metric, public url: string, public apiKey: string) {
+    if (metric === undefined) {
+      throw new Error(`Missing field "metric"`);
+    }
+    if (url === undefined) {
+      throw new Error(`Missing field "url"`);
+    }
+    if (apiKey === undefined) {
+      throw new Error(`Missing field "apiKey"`);
+    }
+  };
+
+  async queryByMetric(from: number, to: number): Promise<{ values: [number, number][], columns: string[] }> {
+
+    if (from > to) {
+      throw new BadRange(
+        `Data-kit got wrong range: from ${from} > to ${to}`,
+        this.metric.datasource.type,
```

Comment:
I think you should not set this `this.metric.datasource.type` because caller oh this method should know which datasource is used already. So maybe you need to catch this error ourside and throw a higher level error with setting this datasoure type and url